### PR TITLE
introducing ios_bfd_templates module

### PR DESCRIPTION
**Pull request by komaldesai13**
## Description
we are introducting ios_bfd_templates module to support  bfd template commands in cisco.ios as per customer requirement

this module supports below commands

```python
bfd-template single-hop BFD300
interval min-tx 300 min-rx 300 multiplier 3
dampening 10 450 2000 54
echo
bfd-template single-hop BFD50
interval min-tx 50 min-rx 50 multiplier 3
dampening 10 450 2000 54
echo
bfd-template multi-hop MH300
interval min-tx 300 min-rx 300 multiplier 3
``` 
## Type of Change

- [x] New module (adding a new module to the collection)

## Related Issue
[ACA-4601](https://issues.redhat.com/browse/ACA-4601)

## Component Name
ios_bfd_templates

## Self-Review Checklist

- [x] I have performed a self-review of my code
- [x] I have added relevant comments to complex code sections
- [x] I have removed all commented code (no commented code should be merged)
- [x] I have updated documentation where needed
- [x] I have considered the security impact of these changes
- [x] I have considered performance implications
- [x] I have thought about error handling and edge cases
- [x] I have tested the changes in my local environment
- [ ] I have verified the changes work with the target IOS version(s)
- [x] I have reviewed the acceptance criteria for related tickets

### Prerequisites

- IOS Version:
- Hardware Platform (if applicable):
- Test Topology:

### Steps to Test
1.
2.
3.

### Test Results

```
```

## Acceptance Criteria Verification

- [x] All acceptance criteria have been reviewed and verified
- [ ] Acceptance criteria checklist:
  - [ ]
  - [ ]
  - [ ]

### Command Output / Logs

```
<!-- Paste below -->
```

### Required Actions

- [x] Requires documentation updates
  
- [x] Requires changelog fragment
  
- [x] Requires integration test updates
- [x] Requires unit test updates
- [ ] Requires coordination with other teams
- [ ] Blocked by PR/MR: #XXX

### CODE REVIEW

File: plugins/module_utils/network/ios/argspec/bfd_templates/bfd_templates.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+    argument_spec = {
+        "config": {
+            "type": "list",
+            "elements": "dict",
+            "options": {
+                "name": {"required": True, "type": "str"},
+                "hop": {
+                    "required": True,
+                    "choices": ["single-hop", "multi-hop"],
```
Comment (KB-perByte) [inline plugins/module_utils/network/ios/argspec/bfd_templates/bfd_templates.py]:
```suggestion
                    "choices": ["single_hop", "multi_hop"],
```

File: plugins/module_utils/network/ios/argspec/bfd_templates/bfd_templates.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+                    "required": False,
+                    "type": "dict",
+                    "options": {
+                        "min_tx": {"required": True, "type": "int"},
+                        "min_rx": {"required": True, "type": "int"},
+                        "multiplier": {"required": True, "type": "int"},
+                    },
+                },
+                "dampening": {
+                    "required": False,
```
Comment (KB-perByte) [inline plugins/module_utils/network/ios/argspec/bfd_templates/bfd_templates.py]: The top level false and nested true for required does not work
This would consider as Required True by default.
Make it free from required attributes

^ but give this a try first

File: plugins/module_utils/network/ios/argspec/bfd_templates/bfd_templates.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+                    "type": "dict",
+                    "options": {
+                        "half_life_period": {"required": True, "type": "int"},
+                        "reuse_threshold": {"required": True, "type": "int"},
```
Comment (KB-perByte) [inline plugins/module_utils/network/ios/argspec/bfd_templates/bfd_templates.py]: These are not required attribute

```
testios(config)#bfd-template multi-hop tests ?
  <cr>  <cr>

testios(config)#bfd-template multi-hop tests 
testios(config-bfd)#?
BFD template configuration commands:
  authentication  Authentication type
  dampening       Enable session dampening
  default         Set a command to its defaults
  echo            Use echo adjunct as bfd detection mechanism
  exit            Exit from BFD template configuration mode
  interval        Transmit interval between BFD packets
  no              Negate a command or set its defaults

testios(config-bfd)#damp
testios(config-bfd)#dampening ?
  <1-30>  Half-life time for the penalty
  <cr>    <cr>

testios(config-bfd)#dampening 22 
```

File: plugins/module_utils/network/ios/argspec/bfd_templates/bfd_templates.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+                        "suppress_threshold": {"required": True, "type": "int"},
+                        "max_suppress_time": {"required": True, "type": "int"},
+                    },
+                },
+                "echo": {"required": False, "type": "bool"},
```
Comment (KB-perByte) [inline plugins/module_utils/network/ios/argspec/bfd_templates/bfd_templates.py]: Remove required false by default it is false.

File: plugins/module_utils/network/ios/argspec/bfd_templates/bfd_templates.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+                "name": {"required": True, "type": "str"},
+                "hop": {
+                    "required": True,
+                    "choices": ["single-hop", "multi-hop"],
+                    "type": "str",
+                },
+                "interval": {
+                    "required": False,
+                    "type": "dict",
+                    "options": {
```
Comment (KB-perByte) [inline plugins/module_utils/network/ios/argspec/bfd_templates/bfd_templates.py:48]: Same here

File: plugins/module_utils/network/ios/argspec/bfd_templates/bfd_templates.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+class Bfd_templatesArgs(object):  # pylint: disable=R0903
+    """The arg spec for the ios_bfd_templates module"""
+
+    argument_spec = {
+        "config": {
+            "type": "list",
+            "elements": "dict",
+            "options": {
+                "name": {"required": True, "type": "str"},
```
Comment (KB-perByte) [inline plugins/module_utils/network/ios/argspec/bfd_templates/bfd_templates.py:40]:
```
testios(config-bfd)#ec
testios(config-bfd)#echo ?
  <cr>  <cr>

testios(config-bfd)#def  
testios(config-bfd)#default ?
  authentication  Authentication type
  dampening       Enable session dampening
  echo            Use echo adjunct as bfd detection mechanism
  interval        Transmit interval between BFD packets

testios(config-bfd)#auth    
testios(config-bfd)#authentication ?
  md5               Keyed md5
  meticulous-md5    Meticulous Keyed md5
  meticulous-sha-1  Meticulous Keyed sha-1
  sha-1             Keyed sha-1

```

We cover all options like auth and default

File: plugins/module_utils/network/ios/argspec/bfd_templates/bfd_templates.py (hunk 1 above)
Comment (komaldesai13) [reply plugins/module_utils/network/ios/argspec/bfd_templates/bfd_templates.py]: resolved

File: plugins/module_utils/network/ios/rm_templates/bfd_templates.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+                        "max_suppress_time": "{{ max_suppress_time | int }}",
+                    },
+                },
+            },
+        },
+        {
+            "name": "echo",
+            "getval": re.compile(
+                r"""
+                \s+(?P<no>no\s)?echo
```
Comment (KB-perByte) [inline plugins/module_utils/network/ios/rm_templates/bfd_templates.py:103]: I am not sure if this relevant to the template module.
Can the config be no echo in appliance?

File: plugins/module_utils/network/ios/rm_templates/bfd_templates.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+        {
+            "name": "authentication",
+            "getval": re.compile(
+                r"""
+                \s+authentication\s(?P<auth_type>sha-1|md5|meticulous-md5|meticulous-sha-1)
+                \skeychain\s(?P<keychain>\S+)
+                $""",
+                re.VERBOSE,
+            ),
+            "setval": "authentication {{ authentication.type|replace('_', '-') }} "
```
Comment (KB-perByte) [inline plugins/module_utils/network/ios/rm_templates/bfd_templates.py]: Can we simplify this?
why do we need replace operation?

File: plugins/module_utils/network/ios/rm_templates/bfd_templates.py [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+            "name": "name",
+            "getval": re.compile(
+                r"""^bfd-template
+                    \s(?P<hop>single-hop|multi-hop)
+                    \s(?P<name>\S+)
+                    $""",
+                re.VERBOSE,
+            ),
+            "compval": "name",
+            "setval": "bfd-template {{ hop|replace('_', '-') }} {{ name }}",
```
Comment (KB-perByte) [inline plugins/module_utils/network/ios/rm_templates/bfd_templates.py]: Can we do it without replace?

File: tests/integration/targets/ios_bfd_templates/tests/cli/round_trip.yaml [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+  cisco.ios.ios_config:
+    lines:
+      - key 1
+      - key-string bfd_test_key
+    parents: key chain bfd_keychain
+
+- name: Create new keychain for replaced test
+  cisco.ios.ios_config:
+    lines:
+      - key 1
```
Comment (KB-perByte) [inline tests/integration/targets/ios_bfd_templates/tests/cli/round_trip.yaml]: Maybe combine it with the previous task, I am not sure

File: tests/unit/modules/network/ios/test_ios_bfd_templates.py [hunk 11]
```diff
@@ -0,0 +1,423 @@
...
+            "interval min-tx 500 min-rx 500 multiplier 5",
+            "dampening 30 2000 5000 54",
+        ]
+        result = self.execute_module(changed=True)
+        self.assertEqual(result["commands"], commands)
+
+    def test_ios_bfd_templates_merged_idempotent(self):
+        self.execute_show_command.return_value = dedent(
+            """\
+            bfd-template single-hop template1
```
Comment (KB-perByte) [inline tests/unit/modules/network/ios/test_ios_bfd_templates.py:87]: Edge cases are hard to predict with limited pre-config for action states.
let's try to extend the unit cases a bit more.

File: changelogs/fragments/bfd_templates.yaml [hunk 12]
```diff
@@ -0,0 +1,3 @@
+---
+minor_changes:
+  - Adds a new Resource Module `ios_bfd_interfaces` to configure BFD on interfaces.
```
Comment (KB-perByte) [inline changelogs/fragments/bfd_templates.yaml]: hehe, this needs to be updated.

File: plugins/module_utils/network/ios/argspec/bfd_templates/bfd_templates.py [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+                "hop": {
+                    "required": True,
+                    "choices": ["single_hop", "multi_hop"],
+                    "type": "str",
+                },
+                "interval": {
```
Comment (KB-perByte) [inline plugins/module_utils/network/ios/argspec/bfd_templates/bfd_templates.py:46]: The required False is not needed, by default, everything is required false.
This change needs to be added in the Module file first.

**Review by KB-perByte — CHANGES_REQUESTED** [review]

File: plugins/module_utils/network/ios/rm_templates/bfd_templates.py (hunk 7 above)
Comment (komaldesai13) [reply plugins/module_utils/network/ios/rm_templates/bfd_templates.py]: "no echo" commands works when we want to use state like override , it basically removes existing echo , see refrence


```python
cisco(config)# bfd-template single-hop template1
cisco(config-bfd)#echo
cisco(config-bfd)#end                               
cisco#show running-config | section ^bfd-template
bfd-template single-hop template1
 interval min-tx 200 min-rx 200 multiplier 3
 authentication sha-1 keychain bfd_keychain
 echo
bfd-template multi-hop template2
 interval min-tx 500 min-rx 500 multiplier 5
 dampening 30 2000 5000 54
cisco#conf t                                     
Enter configuration commands, one per line.  End with CNTL/Z.
cisco(config)# bfd-template single-hop template1
cisco(config-bfd)#no echo
cisco(config-bfd)#end                               
cisco#show running-config | section ^bfd-template
bfd-template single-hop template1
 interval min-tx 200 min-rx 200 multiplier 3
 authentication sha-1 keychain bfd_keychain
bfd-template multi-hop template2
 interval min-tx 500 min-rx 500 multiplier 5
 dampening 30 2000 5000 54
cisco#
```

File: plugins/module_utils/network/ios/rm_templates/bfd_templates.py (hunk 9 above)
Comment (komaldesai13) [reply plugins/module_utils/network/ios/rm_templates/bfd_templates.py]: i found replace is safest option here , string sub method creating complication in code

File: plugins/module_utils/network/ios/argspec/bfd_templates/bfd_templates.py [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+                        "max_suppress_time": {"type": "int"},
+                    },
+                },
+                "echo": {"type": "bool"},
+                "authentication": {
+                    "type": "dict",
+                    "options": {
+                        "type": {
+                            "choices": [
+                                "sha_1",
```
Comment (KB-perByte) [inline plugins/module_utils/network/ios/argspec/bfd_templates/bfd_templates.py:69]: if we are using choice the choices are mostly exact match does sha_1 occurs as is or sha-1, if sha-1 then that should be part of choice, the - to _ change only applies for attributes.

Comment (komaldesai13) [reply plugins/module_utils/network/ios/argspec/bfd_templates/bfd_templates.py]: done , added if conditions
